File: Android/source-code/AppStore/app/src/main/java/org/buildmlearn/appstore/utils/AppReader.java
package org.buildmlearn.appstore.utils;

import android.content.Context;
import android.content.Intent;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import org.buildmlearn.appstore.activities.HomeActivity;
import org.buildmlearn.appstore.models.AppInfo;
import org.buildmlearn.appstore.models.Apps;
import org.buildmlearn.appstore.models.Card;
import org.buildmlearn.appstore.models.FlashModel;
import org.buildmlearn.appstore.models.InfoModel;
import org.buildmlearn.appstore.models.Question;
import org.buildmlearn.appstore.models.QuizModel;
import org.buildmlearn.appstore.models.SpellingsModel;
import org.buildmlearn.appstore.models.WordModel;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Created by Srujan Jha on 6/6/2015.
 */

public class AppReader {
    private static BufferedReader br;

    public static ArrayList<AppInfo> listApps(Context context) {

        ArrayList<AppInfo> mFileList = new ArrayList<AppInfo>();
        Resources res = context.getResources();
        AssetManager am = res.getAssets();
        String appList[],iconList[];
        try {
            appList = am.list("Apps");
            iconList = am.list("Icons");
            if (appList != null) {
                for (int i = 0; i < appList.length; i++) {
                    Log.d("", appList[i]);
                    AppInfo app=new AppInfo();
                    app.Name=(appList[i].substring(0,appList[i].indexOf(".buildmlearn")));
                    app.AppIcon=BitmapFactory.decodeStream(am.open("Icons/" + iconList[i]));
                    BufferedReader br = new BufferedReader(new InputStreamReader(context.getAssets().open("Apps/"+appList[i])));
                    String type=br.readLine();
                    if(type.contains("InfoTemplate"))app.Type=0;
                    else if(type.contains("QuizTemplate"))app.Type=2;
                    else if(type.contains("FlashCardsTemplate"))app.Type=1;
                    else if(type.contains("SpellingTemplate"))app.Type=3;

                    mFileList.add(app);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return mFileList;
    }

    public static String ReadFile(Context myContext, String file) {

        BufferedReader br = null;

        try {
            br = new BufferedReader(new InputStreamReader(myContext.getAssets()
                    .open(file))); // throwing a FileNotFoundException?
            String text = "";
            String tmp = "";
            while ((tmp = br.readLine()) != null) {
                text = text + tmp;
            }
            return text;
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                br.close(); // stop reading
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
        return "";
    }

    public static void readInfoFile(Context myContext, String fileName) {
        InfoModel model = InfoModel.getInstance();
        ArrayList<String> stringList = new ArrayList<String>();
        try {
            HashMap<String, String> mInfoMap = new HashMap<String, String>();
            XMLParser parser = new XMLParser();
            br = new BufferedReader(new InputStreamReader(myContext.getAssets().open(fileName))); // throwing a FileNotFoundException
            String xml = "", temp = "";
            while ((temp = br.readLine()) != null) {xml+=temp;} //getting xml
            Document doc = parser.getDomElement(xml); // getting DOM element
            Element elementAuthor= (Element) doc.getElementsByTagName("author").item(0);
            model.setInfoAuthor(parser.getValue(elementAuthor, "name"));
            model.setInfoName(fileName.substring(5,fileName.length()-12));
            NodeList nodeList = doc.getElementsByTagName("item");
            System.out.println("InfoName"+model.getInfoName()+model.getInfoAuthor());
                // looping through all item nodes <app>
            for (int i = 0; i < nodeList.getLength(); i++) {
                Element elementInfo = (Element) nodeList.item(i);
                String infoKey = parser.getValue(elementInfo, "item_title");
                String infoValue = parser.getValue(elementInfo, "item_description");
                mInfoMap.put(infoKey, infoValue);
                stringList.add(infoKey);
            }
            model.setInfoMap(mInfoMap);
            model.setListTitleList(stringList);
        } catch (Exception e) {
            e.printStackTrace();
        }
        finally {
                try {
                    br.close(); // stop reading
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
        }
    }

    public static void readQuizFile(Context myContext, String fileName) {
        QuizModel model = QuizModel.getInstance();
        ArrayList<Question> mQuestionList = new ArrayList<Question>();

        try {
            XMLParser parser = new XMLParser();

            br = new BufferedReader(new InputStreamReader(myContext.getAssets().open(fileName))); // throwing a FileNotFoundException
            String xml = "", temp = "";
            while ((temp = br.readLine()) != null) {
                xml += temp;
            } //getting xml
            Document doc = parser.getDomElement(xml); // getting DOM element
            Element elementAuthor = (Element) doc.getElementsByTagName("author").item(0);
            model.setQuizAuthor(parser.getValue(elementAuthor, "name"));
            model.setQuizName(fileName.substring(5, fileName.length() - 12));
            NodeList nodeList = doc.getElementsByTagName("item");
            NodeList optionList=doc.getElementsByTagName("option");
            int index=0;
            // looping through all item nodes <app>
            for (int i = 0; i < nodeList.getLength(); i++) {

                Element elementInfo = (Element) nodeList.item(i);
                ArrayList<String> mOptionList = new ArrayList<String>();
                Question ques = new Question();
                ques.setQuestion(parser.getValue(elementInfo, "question"));

                for (int j = 0; j < 4; j++) {
                    mOptionList.add(parser.getElementValue(optionList.item(index++)));
                }
                ques.setAnswerOption(mOptionList);
                ques.setOptionNumber(Integer.parseInt(parser.getValue(elementInfo, "answer")));
                mQuestionList.add(ques);
            }
            model.setQueAnsList(mQuestionList);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                br.close(); // stop reading
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }

    public static void readSpellingsContent(Context myContext, String fileName) {

        SpellingsModel model = SpellingsModel.getInstance();
        ArrayList<WordModel> wordList = new ArrayList<WordModel>();
        try {
            XMLParser parser = new XMLParser();
            br = new BufferedReader(new InputStreamReader(myContext.getAssets().open(fileName))); // throwing a FileNotFoundException
            String xml = "", temp = "";
            while ((temp = br.readLine()) != null) {
                xml += temp;
            } //getting xml
            Document doc = parser.getDomElement(xml); // getting DOM element
            Element elementAuthor = (Element) doc.getElementsByTagName("author").item(0);
            model.setPuzzleAuthor(parser.getValue(elementAuthor, "name"));
            model.setPuzzleName(fileName.substring(5, fileName.length() - 12));
            NodeList nodeList = doc.getElementsByTagName("item");
            // looping through all item nodes <app>
            for (int i = 0; i < nodeList.getLength(); i++) {

                Element elementInfo = (Element) nodeList.item(i);
                WordModel word = new WordModel(
                        parser.getValue(elementInfo, "word"),
                        parser.getValue(elementInfo, "meaning"));
                wordList.add(word);
            }
            model.setSpellingsList(wordList);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                br.close(); // stop reading
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }

    public static void readFlashContent(Context myContext, String fileName) {

        FlashModel model = FlashModel.getInstance();
        ArrayList<Card> cardList = new ArrayList<Card>();
        try {
            XMLParser parser = new XMLParser();
            br = new BufferedReader(new InputStreamReader(myContext.getAssets().open(fileName))); // throwing a FileNotFoundException
            String xml = "", temp = "";
            while ((temp = br.readLine()) != null) {
                xml += temp;
            } //getting xml
            Document doc = parser.getDomElement(xml); // getting DOM element
            Element elementAuthor = (Element) doc.getElementsByTagName("author").item(0);
            model.setAuthor(parser.getValue(elementAuthor, "name"));
            model.setName(fileName.substring(5, fileName.length() - 12));
            NodeList nodeList = doc.getElementsByTagName("item");
            // looping through all item nodes <app>
            for (int i = 0; i < nodeList.getLength(); i++) {

                Element elementInfo = (Element) nodeList.item(i);
                Card card = new Card(
                        parser.getValue(elementInfo, "question"),
                        parser.getValue(elementInfo, "answer"),
                        parser.getValue(elementInfo, "hint"),
                        parser.getValue(elementInfo, "image"));
                cardList.add(card);
            }
            model.setList(cardList);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                br.close(); // stop reading
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }
}

File: WP/source-code/AppStore/AppStore/Common/AppReader.cs
using AppStore.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using Windows.Data.Xml.Dom;
using Windows.Storage;

namespace AppStore.Common
{
    class AppReader
    {
        /*
        public static List<Apps> listApps()
        {
            List<AppInfo> mFileList = new List<AppInfo>();
            SharedPreferences SP = PreferenceManager.getDefaultSharedPreferences(context);
            Resources res = context.getResources();
            AssetManager am = res.getAssets();
            string appList[], iconList[];
            try
            {
                appList = am.list("Apps");
                iconList = am.list("Icons");
                if (appList != null)
                {
                    for (int i = 0; i < appList.length; i++)
                    {
                        AppInfo app = new AppInfo();
                        app.Name = (appList[i].substring(0, appList[i].indexOf(".buildmlearn")));
                        if (!SP.contains(app.Name))
                        {
                            SharedPreferences.Editor editor1 = SP.edit();
                            editor1.putBoolean(app.Name, false);
                            editor1.commit(); continue;
                        }
                        if (!SP.getBoolean(app.Name, false)) continue;

                        app.AppIcon = BitmapFactory.decodeStream(am.open("Icons/" + iconList[i]));
                        BufferedReader br = new BufferedReader(new InputStreamReader(context.getAssets().open("Apps/" + appList[i])));
                        string type = br.readLine();
                        if (type.contains("InfoTemplate")) app.Type = 0;
                        else if (type.contains("QuizTemplate")) app.Type = 2;
                        else if (type.contains("FlashCardsTemplate")) app.Type = 1;
                        else if (type.contains("SpellingTemplate")) app.Type = 3;
                        br.readLine();
                        type = br.readLine();
                        int x = type.indexOf("<name>") + 6;
                        int y = type.indexOf("<", x + 1);
                        app.Author = type.substring(x, y);
                        mFileList.add(app);
                    }
                }
            }
            catch (IOException e)
            {
                e.printStackTrace();
            }
            AppList = mFileList;
            return mFileList;
        }
        */
        public static void readInfoFile(string fileName)
        {
            try
            {
                InfoModel model = InfoModel.getInstance();
                List<string> infoTitleList = new List<string>();
                List<string> infoDescriptionList = new List<string>();
                XmlDocument doc = new XmlDocument();
                doc.LoadXml(XDocument.Load("Assets/" + fileName + ".xml").ToString());
                model.setInfoName(doc.GetElementsByTagName("title").ElementAt(0).InnerText.Trim());
                model.setInfoDescription(doc.GetElementsByTagName("description").ElementAt(0).InnerText.Trim());
                string[] author = doc.GetElementsByTagName("author").ElementAt(0).InnerText.Split('\n');
                model.setInfoAuthor(author[1].Trim());
                model.setInfoAuthorEmail(author[2].Trim());
                model.setInfoVersion(doc.GetElementsByTagName("version").ElementAt(0).InnerText.Trim());
                XmlNodeList info_title = doc.GetElementsByTagName("item_title");
                XmlNodeList info_description = doc.GetElementsByTagName("item_description");
                for (int i = 0; i < info_title.Length; i++)
                {
                    infoTitleList.Add(info_title.ElementAt(i).InnerText.Trim());
                    infoDescriptionList.Add(info_description.ElementAt(i).InnerText.Trim());
                }
                model.setListTitleList(infoTitleList);
                model.setInfoDescriptionList(infoDescriptionList);
            }
            catch (Exception) { }
        }
        public static void readQuizFile(string fileName)
        {
            try
            {
                QuizModel model = QuizModel.getInstance();
                List<Question> mQuestionList = new List<Question>();
                XmlDocument doc = new XmlDocument();
                doc.LoadXml(XDocument.Load("Assets/" + fileName + ".xml").ToString());
                model.setQuizName(doc.GetElementsByTagName("title").ElementAt(0).InnerText.Trim());
                model.setQuizDescription(doc.GetElementsByTagName("description").ElementAt(0).InnerText.Trim());
                string[] author = doc.GetElementsByTagName("author").ElementAt(0).InnerText.Split('\n');
                model.setQuizAuthor(author[1].Trim());
                model.setQuizAuthorEmail(author[2].Trim());
                model.setQuizVersion(doc.GetElementsByTagName("version").ElementAt(0).InnerText.Trim());
                XmlNodeList questions = doc.GetElementsByTagName("item");
                for (int i = 0; i < questions.Length; i++)
                {
                    Question q = new Question();
                    string[] ar = questions.ElementAt(i).InnerText.Split('\n');
                    List<string> options = new List<string>();
                    options.Add(ar[2].Trim());
                    options.Add(ar[3].Trim());
                    options.Add(ar[4].Trim());
                    options.Add(ar[5].Trim());
                    q.setOptionNumber(Convert.ToInt16(ar[6].Trim()));
                    q.setAnswerOption(options);
                    q.setQuestion(ar[1].Trim());
                    mQuestionList.Add(q);
                }
                model.setQueAnsList(mQuestionList);
            }
            catch (Exception) { }
        }
        public static void readSpellingsFile(string fileName)
        {
            try
            {
                SpellingsModel model = SpellingsModel.getInstance();
                List<WordModel> wordList = new List<WordModel>();
                XmlDocument doc = new XmlDocument();
                doc.LoadXml(XDocument.Load("Assets/" + fileName + ".xml").ToString());
                model.setPuzzleName(doc.GetElementsByTagName("title").ElementAt(0).InnerText.Trim());
                model.setPuzzleDescription(doc.GetElementsByTagName("description").ElementAt(0).InnerText.Trim());
                string[] author = doc.GetElementsByTagName("author").ElementAt(0).InnerText.Split('\n');
                model.setPuzzleAuthor(author[1].Trim());
                model.setPuzzleAuthorEmail(author[2].Trim());
                model.setPuzzleVersion(doc.GetElementsByTagName("version").ElementAt(0).InnerText.Trim());
                XmlNodeList item = doc.GetElementsByTagName("item");
                // looping through all item nodes <app>
                for (int i = 0; i < item.Length; i++)
                {
                    string[] ar = item.ElementAt(i).InnerText.Split('\n');
                    WordModel word = new WordModel(ar[1].Trim(), ar[2].Trim());
                    wordList.Add(word);
                }
                model.setSpellingsList(wordList);
            }
            catch (Exception) { }
        }
        public static void readFlashFile(string fileName)
        {
            try
            {
                FlashModel model = FlashModel.getInstance();
                List<Card> cardList = new List<Card>();
                XmlDocument doc = new XmlDocument();
                doc.LoadXml(XDocument.Load("Assets/Apps/" + fileName + ".xml").ToString());
                model.setFlashName(doc.GetElementsByTagName("title").ElementAt(0).InnerText.Trim());
                model.setFlashDescription(doc.GetElementsByTagName("description").ElementAt(0).InnerText.Trim());
                string[] author = doc.GetElementsByTagName("author").ElementAt(0).InnerText.Split('\n');
                model.setFlashAuthor(author[1].Trim());
                model.setFlashAuthorEmail(author[2].Trim());
                model.setFlashVersion(doc.GetElementsByTagName("version").ElementAt(0).InnerText.Trim());
                XmlNodeList item = doc.GetElementsByTagName("item");
                // looping through all item nodes <app>
                for (int i = 0; i < item.Length; i++)
                {
                    string[] ar = item.ElementAt(i).InnerText.Split('\n');
                    Card card = new Card(ar[1], ar[2], ar[3], ar[4]);
                    cardList.Add(card);
                }
                model.setCardList(cardList);
            }
            catch (Exception) { }
        }
    }
}
